### PR TITLE
feat(server-metrics): warmup-phase export, realtime robustness, per-worker logging

### PR DESCRIPTION
**Pull request by ajcasagrande**
## Summary

Server-metrics improvements, standalone (no dependency on the agentx feature branch):

- **Warmup metrics export (schema 1.1):** scrape records stamp `benchmark_phase`; `export_results` aggregates a separate warmup window and emits `warmup_metrics` alongside the profiling-only top-level `metrics`, with `phase_time_ranges` in the summary. Top-level `metrics` stays profiling-only for backward compatibility (`metrics_phase` pins this for readers). Warmup window fields follow the NaN/Inf discipline (finite or None).
- **Realtime/robustness:** `get_indices_for_filter` returns an explicit `None` final index instead of clamping; histogram series-length mismatches warn once instead of log-spamming; realtime counter resets no longer poison cumulative deltas.
- **Collectors:** per-worker logging and a GPU realtime guard.
- **parquet metadata:** reads the v2 phase field `rate` (also fixed independently in #1093 — identical one-liner, merges cleanly in either order).

Docs: server-metrics JSON schema doc updated for 1.1.

## Testing

- New adversarial suites: realtime counter reset, SGLang residuals, plus warmup-export coverage in the accumulator/json-exporter/manager tests.
- `pytest tests/unit/server_metrics/ tests/unit/gpu_telemetry/ tests/unit/records/` — 1156 passed.

## Summary by CodeRabbit

* **New Features**
  * Added warmup-aware server-metrics exports (schema **1.1**), including `metrics_phase`, `phase_time_ranges`, and `warmup_metrics`, plus warmup endpoint summaries and phase-tagged benchmark records.
  * Added time-window querying (`query_time_range`) and stream-style lifecycle hooks (`process_record`/`finalize`) for JSONL telemetry export.
  * Added `realtime_snapshot()` and UI-type-aware realtime interval handling (interval `0` disables reporting).
* **Bug Fixes**
  * Improved histogram and time-window export guards, including empty/invalid-window handling and safer end-of-collection behavior.
* **Documentation**
  * Updated server-metrics JSON schema docs and clarified realtime metrics interval environment variable semantics.
* **Chores**
  * Expanded unit tests for warmup, realtime, and edge cases.

### CODE REVIEW

File: src/aiperf/common/models/server_metrics_models.py [hunk 1]
```diff
@@ -659,6 +679,19 @@ class ServerMetricsExportData(AIPerfBaseModel):
...
+                str,
+                GaugeMetricData | CounterMetricData | HistogramMetricData,
+            ]
+        ]
+        | None
+    ) = Field(
+        default=None,
+        description="Server metrics aggregated only over the warmup phase. "
+        "Absent when no warmup phase ran or no warmup server metrics were collected.",
+    )
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/common/models/server_metrics_models.py:698]: Fixed in 942c6a70e: Widened ServerMetricsExportData.warmup_metrics value union in src/aiperf/common/models/server_metrics_models.py to GaugeMetricData | CounterMetricData | HistogramMetricData | UnknownMetricData, matching the profiling metrics field, so warmup exports containing untyped Prometheus families no longer raise pydantic ValidationError. Added regression test test_export_data_accepts_unknown_in_warmup_metrics in tests/unit/server_metrics/test_unknown_metrics.py.

File: src/aiperf/server_metrics/accumulator.py [hunk 2]
```diff
@@ -131,23 +153,46 @@ async def export_results(
...
+            self._slice_duration,
+            include_final_collection=True,
         )
+        warmup_endpoint_summaries = None
+        if warmup_start_ns is not None and warmup_end_ns is not None:
+            warmup_endpoint_summaries = self._compute_endpoint_summaries(
+                warmup_start_ns,
+                warmup_end_ns,
+                self._slice_duration,
+                include_final_collection=False,
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/server_metrics/accumulator.py:198]: Fixed in 942c6a70e: ServerMetricsAccumulator (src/aiperf/server_metrics/accumulator.py) now tracks the latest WARMUP-tagged record timestamp (_last_warmup_record_ns, set in process_server_metrics_record) and in export_results extends the warmup summary window end to max(warmup_end_ns, min(_last_warmup_record_ns, start_ns - 1)), keeping include_final_collection=False. The dedicated end-of-warmup scrape (which lands strictly after warmup_end_ns) is now included in warmup counter/histogram deltas while the profiling-start cap guarantees profiling samples are never admitted. Added test_export_results_warmup_includes_final_warmup_scrape in tests/unit/server_metrics/test_accumulator.py asserting the warmup delta includes the post-warmup_end scrape (100 vs the old 50) and the profiling delta is unchanged.

File: src/aiperf/common/environment.py [hunk 3]
```diff
@@ -1132,16 +1132,27 @@ class _UISettings(BaseSettings):
...
         default=False,
         description="Enable real-time metrics collection and reporting despite UI type",
     )
+
+    def realtime_metrics_interval(self, ui_type: UIType) -> float:
+        """Resolve the realtime metrics tick interval, applying the auto-default by UI type."""
+        if self.REALTIME_METRICS_INTERVAL is not None:
+            return self.REALTIME_METRICS_INTERVAL
+        return 5.0 if ui_type == UIType.DASHBOARD else 30.0
+
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/common/environment.py:1212]: Fixed in 942c6a70e: _report_realtime_inference_metrics_task in src/aiperf/records/records_manager.py now resolves interval = Environment.UI.realtime_metrics_interval(self.run.cfg.runtime.ui) before the loop, returns early when interval == 0 (documented disable semantics, avoids asyncio.sleep(0) busy-spin), and sleeps on the resolved interval — mirroring the GPU telemetry accumulator pattern and eliminating the TypeError when REALTIME_METRICS_INTERVAL is None.

File: src/aiperf/server_metrics/accumulator.py [hunk 4]
```diff
@@ -340,3 +392,434 @@ async def summarize(self) -> list[MetricResult]:
             Empty list (server metrics exported via export_results instead)
         """
         return []
+
+    def realtime_snapshot(self, start_ns: int | None = None) -> dict[str, float]:
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/server_metrics/accumulator.py:613]: Correct that nothing in `src/` calls `realtime_snapshot()` at this PR's head - that is a deliberate staged landing (this PR is an extraction slice from the agentx branch, with 'Realtime snapshot' as a listed bullet). The consumer already exists on the parent branch (`records_manager.py` renders it into the realtime log block with its own delta bookkeeping) and lands in the next slice. Wiring a quick `MetricResult` integration here would inject unregistered metric tags into `RealtimeMetricsMessage` and diverge from that design.

**Comment by matthewkotila** [comment]
Looks great overall--nice set of fixes.

**1.** 🟠 **Unwired AgentX scaffolding ships with no production callers.** `realtime_snapshot()` + its ~10 helpers, `query_time_range()`/`_timestamps_ns`, and the `process_record()`/`finalize()` aliases are referenced only by tests — nothing in `src/` calls them. If that's intended groundwork, fine, but two nits regardless: the alias docstrings cite `AccumulatorProtocol` / `StreamExporterProtocol`, which don't exist (the real ones are `ServerMetricsAccumulatorProtocol` / `GPUTelemetryAccumulatorProtocol`, and the only production `process_record` is a different 2-arg interface). Worth correcting the docstrings, and deciding whether this lands here or with the PR that wires it.

**2.** 🟡 **`_to_pct` can inflate a real percentage 100× (latent).** On the last-resort SGLang `cache_hit_rate` gauge fallback, a genuine 0.8% rate is `<= 1.0` and gets multiplied to 80%. The 0–1 vs 0–100 ambiguity is unknowable from one sample. Only reachable once `realtime_snapshot` is wired, but the tests bake it in.

**3.** 🟡 **`query_time_range` uses half-open `[start, end)` while the rest of the module is inclusive (latent).** `get_time_mask`/`get_indices_for_filter` include the sample at exactly `end_ns` (`side="right"`); this mask drops it, so the analyzer and the aggregation path can disagree on a boundary sample. Worth aligning the convention before it's consumed.

**4.** 🟡 **Warmup final-scrape inclusion silently degrades at zero gap.** The `min(self._last_warmup_record_ns, start_ns - 1)` cap correctly prevents profiling samples leaking into warmup — but if profiling `start_ns == warmup_end_ns` (e.g. a seamless warmup), the cap can fall below `warmup_end_ns` and the "include the dedicated end-of-warmup scrape" benefit just vanishes with no warning. A phase-tag-based window (the `benchmark_phase` field is already on the record but only used to compute `_last_warmup_record_ns`) would be more robust than the timestamp arithmetic.

🟢 **Minor:**
- `manager.py` `record.model_copy(update={"benchmark_phase": ...})` per scrape — these records are freshly built and unshared, so direct assignment avoids the clone.
- `json_exporter.py` keys `phase_time_ranges` with bare strings `"profiling"`/`"warmup"` while the sibling `metrics_phase` is a `CreditPhase` enum — two spellings that can drift.
- `query_time_range`/`_timestamps_ns` setup and the `interval == 0` guard are duplicated across the two accumulators / realtime tasks — candidates for a shared mixin / `_UISettings` helper.

Nothing blocking on the paths that are actually exercised.

**Comment by ajcasagrande** [comment]
Thanks for the thorough pass — agreed nothing is blocking on the exercised paths. Re-verified each point against current HEAD and against `ajc/agentx-full-port`:

**1. Unwired scaffolding** — no longer accurate on this branch. `process_record` / `finalize` are live via `AccumulatorProtocol` / `StreamExporterProtocol` in `aiperf.common.accumulator_protocols` (those protocols do exist; earlier wording about them being nonexistent was wrong). `realtime_snapshot` is already consumed by `RecordsManager._collect_realtime_server_snapshot`. Same wiring is present on agentx-full-port.

**2. `_to_pct` 100× inflation** — the ambiguity is real in the abstract, but the cited SGLang `cache_hit_rate` path is documented as a **ratio** (0–1), so `0.8 → 80%` is correct under that contract. A server that encoded a sub-1% reading as an already-percent value in `(0, 1]` cannot be distinguished from an 80% ratio. Left the heuristic as-is and documented the contract on `_to_pct`; counter-pair sources remain preferred. Same code on agentx-full-port.

**3. `query_time_range` half-open vs inclusive** — intentional. Half-open `[start, end)` matches `AccumulatorProtocol` and the metrics accumulator. Per-series `get_time_mask` / `get_indices_for_filter` are a different API (inclusive). Clarified in the accumulator docstrings rather than changing the protocol. Same on agentx-full-port.

**4. Warmup final-scrape at zero gap** — already fixed here via phase-tag aggregation (`_compute_phase_endpoint_summaries(CreditPhase.WARMUP, …)`). agentx-full-port still uses the older `_last_warmup_record_ns` / `start_ns - 1` cap.

**Minors:**
- `model_copy` → already direct `record.benchmark_phase = …` on this PR (agentx still clones).
- `phase_time_ranges` keys → already `CreditPhase.PROFILING` / `WARMUP` (agentx still uses bare strings).
- Shared mixin for `query_time_range` / interval-0 guard → defer; fine as follow-up once both consumers stay stable.

Also landed a small docs pass: schema minimal example → 1.1 + `metrics_phase`, documented `phase_time_ranges` in the Summary section, and refreshed the adversarial realtime docstring for the clamped-delta behavior.

File: src/aiperf/server_metrics/accumulator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/server_metrics/accumulator.py:613]: Update after re-verification: this is no longer accurate on current HEAD. `RecordsManager._collect_realtime_server_snapshot` already calls `realtime_snapshot` (same on agentx-full-port). Resolving.

**Review by matthewkotila — APPROVED** [review]
